### DISCUSSION TITLE
App mounting vs log replace

### DISCUSSION THREAD

**Discussion by daleif** [mojolicious/mojo#2215] [General]
We have several small webapps that don't see much traffic. Thought it might be an idea to serve them all under one hyptotoad as mounted/embedded apps.

This seems to be working just fine. Just one issue:

I'd like to have each mounted app log to their own separate log file. For my standalone HT installations I just use

```
  if ( $self->mode eq 'production' && $config->{'log_file_production'} ) {
      $self->log(
	  Mojo::Log->new(
	      path => $config->{'log_file_production'},
	      level => 'info'
	  ));
  }
```

crude, but it works just fine.

Inside `Mojolicious::Plugin::Mount` we see 

```
my $embed = Mojo::Server->new->load_app($conf->{$path})->secrets($app->secrets)->log($app->log);
```
so the `log` obj is being passed on from the parent app. 

The earlier trick above seems to replace the log obj inside the  `startup` method of the mounted app, but not in one of its controllers. I tried dumping the log-obj  one has `handle = stderr`,  `startup` has `path` setting.

Any ideas for  properly replacing the log obj inside each mounted apps?

(if relevant all tests have been made using `hyptotoad -f`, so no testing having it run under systemd)

**Comment by s1037989** [comment]
I think a shared log is by design, and the [t/mojolicious/embedded_lite_app.t](https://github.com/mojolicious/mojo/blob/main/t/mojolicious/embedded_lite_app.t#L432) test reflects that as well.

From the [Mount Synopsis](https://docs.mojolicious.org/Mojolicious/Plugin/Mount#SYNOPSIS):

In your hypnotoad app:
```
# Adjust the generated route and mounted application
my $example = plugin Mount => {'/example' => '/home/sri/example.pl'};
$example->to(message => 'It works great!');
my $app = $example->pattern->defaults->{app};
$app->config(foo => 'bar');
$app->log(Mojo::Log->new(path => 'example.log'));   #  set your app-specific log here.
```

Alternatively, you could make your own Mount plugin that sets the log path as you want, for example:
```
my $embed = Mojo::Server->new->load_app($conf->{$path})->secrets($app->secrets);
$embed->log->path("$conf->{$path}.log");
```

After all, per the [Mount Description](https://docs.mojolicious.org/Mojolicious/Plugin/Mount#DESCRIPTION):
```
The code of this plugin is a good example for learning to build new plugins, you're welcome to fork it.
```

Both methods will cause all app->log references within a mounted app to log to the specific app-specific log file.

This is not enough for the default helper `log`.  Add `helper log => sub ($c) { $c->app->log };` to each mounted app.  You lose the context of the request id, but I'm sure there's a way to retain that some how.

**Reply by daleif** [reply]
Thanks, will make some tests. The location of each log is in the config of each of the mounted apps, but we can access those...

I just wonder why it is too late to change it properly inside the setup method of the app. I can overwrite the secrets, but that is probably not as complicated as replacing the log inside each.

I currently mount three apps, so will test if we can do this will all of them.

**Reply by daleif** [reply]
Missed the last to sentences. Then it works. Thanks
